### ISSUE TITLE
RWByteAddressBuffer not visible in UAVs list

### ISSUE THREAD

**Issue by sebbbi** [baldurk/renderdoc#342]
Bytebuffers are only visible in compute shader "Pipeline State - Resources" list when bound as SRV (hlsl ByteAddressBuffer). When I bind the same bytebuffer as UAV (hlsl RWByteAddressBuffer), it is no longer visible in the RenderDoc UI at all (it's neither in "Resources" or "UAVs" list).

This bug is reproducible in multiple shaders (all are working properly).

**Comment by baldurk** [comment]
I just checked with a simple compute shader like so with one of my sample tests that used byte address buffers:

```hlsl
ByteAddressBuffer inbuf : register(t0);
RWByteAddressBuffer outbuf : register(u0);

[numthreads(1, 1, 1)]
void main()
{
	uint4 data = inbuf.Load4(5*4);
	outbuf.Store4(10*4, data);

	data.xy = inbuf.Load2(15*4);
	outbuf.Store2(0, data.xy);
}
```

And they displayed mostly correctly (there was a small bug where they were listed as structured buffers instead of byte address - I've fixed that).

![image](https://cloud.githubusercontent.com/assets/661798/18093939/edb86af8-6ed1-11e6-8f03-06f90ad37580.png)

Is it possible that the shader just doesn't reference those UAVs? By default the pipeline view only shows bindings that exist in the shader rather than showing all possible resources that are bound, since for some programs that could be a long list as there's no need to redundantly unbind unused resources.

You can check by clicking "Show Disabled Items" in the top to toggle displaying all non-NULL bindings "Show Empty Items" will show NULL unused bindings too.

**Comment by sebbbi** [comment]
The output buffers are certainly bound correctly, since the results are properly read in the next dispatch (as SRV). If open pipeline state of the next buffer, I can see the input buffer. The buffer debug verifies that data has been properly written.

I am using Unreal Engine with a custom RenderDoc plugin. I need to check whether UE binds the RWByteAddressBuffer to some odd register index, etc.

I will check resulting shader code, etc tomorrow, and give you more info. This bug is occurring for multiple shaders. I have mandatory binding set (in UE). This gives assert if a shader doesn't use all of the bindings (for example the compiler has optimized out a statement).

**Comment by baldurk** [comment]
A-ha! Yes I have repro'd it now, mentioning Unreal gave me the hint. It looks like when you have very aggressive shader stripping that removes reflection data, the UAV is incorrectly marked as unused. If I strip the reflection data from my shaders that's what I get, you can try clicking 'Show disabled' to see if that's the same for you:

![image](https://cloud.githubusercontent.com/assets/661798/18094836/22097114-6ed5-11e6-90a8-7ada84e1beb2.png)

I'll fix it now, probably just one line missing in my reflection reconstruction.

btw though I'd definitely recommend getting unstripped shaders into renderdoc somehow - it _really_ significantly improves the experience of debugging. I don't know UE4 myself much but talking to a friend who does he suggests wiping your shader cache and setting these cvars:

```r.ShaderDevelopmentMode=1
r.DumpShaderDebugInfo=1
r.DumpShaderDebugShortNames=1
r.Shaders.Optimize=0
r.Shaders.KeepDebugInfo=1
```

With more information on them in `ConsoleVariables.ini`.

**Comment by baldurk** [comment]
If the bug is as described above, I think that commit should fix it. Let me know how you get on, if "Show Disabled" makes it show up for you then hopefully the next nightly build will work.

**Comment by sebbbi** [comment]
Thanks a lot for the quick response and quick bugfix!

I confirmed that: A) "Show disabled" shows the RW buffer binding correctly. B) only stripped disassembly of shaders is loaded. C) constant buffer declarations are just float4 (have to manually insert the buffer layout). UE is definitely aggressively stripping every bit of debug info.

I am relatively new to UE. Have been developing a custom graphics engine at my previous job. RenderDoc worked fine with that engine. Will try these config flags to get more debug output from UE shaders. Tried to find some documentation about the matter, but the UE plugin (https://github.com/Temaran/UE4RenderDocPlugin) has really bare bone docs. This topic is not even mentioned.

**Comment by sebbbi** [comment]
UE wiki page about shader debugging with RenderDoc: https://wiki.unrealengine.com/Debugging_Unreal_Shader_Files

**Comment by baldurk** [comment]
Yeh I'm afraid I don't really know much more about UE4 and I can't help with that side of things. I thought maybe the plugin you're using had an option to enable shader debug info but I've not used it myself.

RenderDoc does allow user code to split apart the unstripped 'debug' blobs and save them separately from the actual stripped shaders, and then just keep an identifier in the stripped blob to look it up - just like .exes and .pdbs. Unfortunately this isn't part of the API so it needs the code to implement it.

**Comment by sebbbi** [comment]
Investigated UE shader debugging a bit more and added the flags you suggested. Now the names of inputs and constant buffer variables show correctly. The missing RWByteAddressBuffer also shows properly (but it is shown as StructuredBuffer like you said before). This is a big improvement.

Shader source code is also now shown (previously only disassembly was shown). UE is merging all shader includes together in one big hlsl file. The big hlsl file is shown properly. Disassembly now also shows line numbers (to hlsl), but the line numbers seem to point to the original (non-merged) file and the path is also wrong. Still it's a huge improvement compared how it worked previously.

Editing shaders in capture now also works perfectly. This is a major improvement to my iteration time, as my shader iteration loop mostly contains repeated shader edit in capture + run timings again.

**Comment by baldurk** [comment]
Yeh unfortunately most of this is on the responsibility of the application so I can't do much from RenderDoc's side to improve it - things like the #line referring to the original file, the shader debug data only embeds the actual file used to compile so I can't do much with just "foo.hlsl" if the actual shader embedded is "generated_merged_output.hlsl".

Did you try with the latest nightly build? The structured buffer bug should be fixed, I sorted that when I first looked into the problem. It will still be broken on v0.30 though.
